homeassistant/components/esphome/__init__.py

```diff
@@ -199,93 +200,213 @@ async def on_login() -> None:
+            if self._entry.entry_id not in self._hass.data[DOMAIN]:
+                # When removing/disconnecting manually
+                return
+            device_registry = (
+                await self._hass.helpers.device_registry.async_get_registry()
+            )
+            devices = dr.async_entries_for_config_entry(
+                device_registry, self._entry.entry_id
+            )
+            for device in devices:
+                # There is only one device in ESPHome
+                if device.disabled:
+                    # Don't attempt to connect if it's disabled
+                    return
 
-    return try_connect
+            await self._try_connect()
+
+    async def start(self):
+        """Start the reconnecting logic background task."""
+        # Create reconnection loop outside of HA's tracked tasks in order
+        # not to delay startup.
+        self._loop_task = self._hass.loop.create_task(self._reconnect_loop())
+        # Listen for mDNS records so we can reconnect directly if a received mDNS record
+        # indicates the node is up again
+        await self._hass.async_add_executor_job(self._zc.add_listener, self, None)
+
+        async with self._connected_lock:
+            self._connected = False
+        self._reconnect_event.set()
+
+    async def stop(self):
+        """Stop the reconnecting logic background task. Does not disconnect the client."""
+        if self._loop_task is not None:
+            self._loop_task.cancel()
+            self._loop_task = None
+        await self._hass.async_add_executor_job(self._zc.remove_listener, self)
+
+    @callback
+    def stop_callback(self):
+        """Stop as an async callback function."""
+        self._hass.async_create_task(self.stop())
+
+    def update_record(self, zc: Zeroconf, now: float, record: DNSRecord) -> None:
+        """Listen to zeroconf updated mDNS records."""
+        if not isinstance(record, DNSPointer):
+            # We only consider PTR records and match using the alias name
+            return
+        if self._entry_data is None or self._entry_data.device_info is None:
+            # Either the entry was already teared down or we haven't received device info yet
+            return
+        filter_alias = f"{self._entry_data.device_info.name}._esphomelib._tcp.local."
+        if record.alias != filter_alias:
+            return
+
+        # This is a mDNS record from the device and could mean it just woke up
+        # Check if already connected, no lock needed for this access
+        if self._connected:
+            return
+
+        # Tell reconnection logic to retry connection attempt now (even before reconnect timer finishes)
+        async def set_reconnect():
+            self._reconnect_event.set()
+
+        _LOGGER.debug(
+            "%s: Triggering reconnect because of received mDNS record %s",
+            self._host,
+            record,
+        )
+        self._hass.add_job(set_reconnect)
```

Comment:
Would be slightly more efficient to call this as `self._set_reconnect` instead of the nested function